beets/dbcore/db.py

```diff
@@ -306,7 +317,7 @@ class Model(ABC):
     """
 
     @cached_classproperty
-    def _relation(cls) -> type[Model]:
+    def _relation(cls):
```

Comment:
I suspect this should really be ` -> Self`? Maybe we could do

```py
if sys.version_info >= (3, 11):
    from typing import Self
else:
    from typing_extensions import Self
```

to make this work? (We already do this in another place: See the `aura` plugin.)

Reply:
Or rather, `type[Self]`.

Reply:
I was hoping it would be possible to express this with `Self`, but there are two issues:

1. This returns a model class, so it's more like `type[Self]` 😅 
2. It returns a sibling class compared with the class that the property is under. If I understood their docs correctly

> ... you can import the special type typing.Self that is automatically transformed into a method-level type parameter with the _current class as the **upper bound**_,

this is probably not suitable for our use case?

Ultimately this is just a generic property definition - both of the models define specific returns types where this property is used:

* `Album`:
```py
    @cached_classproperty
    def _relation(cls) -> type[Item]:
        return Item
```
* `Item`:
```py
    @cached_classproperty
    def _relation(cls) -> type[Album]:
        return Album
```

I was going to `raise NotImplementedError` in the generic `Model._relation` implementation, but I found that this kind of broke the `ModelFixture` model in `test_dbcore`, so I just left it to return `cls`.

Reply:
Oh, I misunderstood what's going on here: Even `type[Self]` is definitely wrong, for the reason you explained. Now I'm confused as to why `type[Model]` as it was before doesn't work: From the [mypy docs](https://mypy.readthedocs.io/en/stable/class_basics.html#overriding-statically-typed-methods)

> You can also vary return types covariantly in overriding. For example, you could override the return type Iterable[int] with a subtype such as list[int]. Similarly, you can vary argument types contravariantly – subclasses can have more general argument types.

But feel free to not address this in this PR, and just get in what we have quickly :)

Reply:
More precisely, I'd say this should be `type[Model[D]]`, but right now, it seems that for me locally, mypy doesn't really check this at all: I don't even manage to make it complain about that line.

Reply:
> Now I'm confused as to why type[Model] as it was before doesn't work: From the [mypy docs](https://mypy.readthedocs.io/en/stable/class_basics.html#overriding-statically-typed-methods)

Now this has to do with the custom descriptor `cached_classproperty` - I have spent far too many hours battling with mypy trying to make it aware that it works the same way like `classmethod` and that `cls` is the **class** and not the object instance. I even learned how to write mypy plugins but I still failed 😅

Try adding the output type and see what happens

```py
    @cached_classproperty
    def _relation(cls) -> type[Model]:
        """The model that this model is closely related to."""
        return cls  # mypy: Incompatible return value type (got "Model[D]", expected "type[Model[Any]]")
```

Reply:
That's what I did — but nothing bad happened... but I'm not going to follow you into that rabbit hole and start investigating this now ;)

Reply:
Best of luck! It would be amazing if you can figure this out.

Reply:
Just thought I'd note that I use it in [`beetcamp`](https://github.com/snejus/beetcamp/blob/main/beetsplug/bandcamp/catalognum.py) slightly differently and it works as expected there:

```py
class cached_classproperty(Generic[T]):
    def __init__(self, getter: Callable[..., T]) -> None:
        self.getter = getter
        self.cache: dict[type[object], T] = {}

    def __get__(self, instance: object, owner: type[object]) -> T:
        if owner not in self.cache:
            self.cache[owner] = self.getter(owner)

        return self.cache[owner]


def cached_patternprop(
    pattern: str, flags: int = 0
) -> cached_classproperty[Pattern[str]]:
    """Pattern is compiled and cached the first time it is accessed."""
    return cached_classproperty(lambda _: re.compile(pattern, flags))


@dataclass
class Catalognum:
    ...
    start_end = cached_patternprop(rf"((^{MATCH})|({MATCH}$))", re.M | re.VERBOSE)
    # enclosed by parens or square brackets, but not ending with MIX
    delimited = cached_patternprop(rf"(?:[\[(])({MATCH})(?:[])]|$)", re.VERBOSE)
```

`Catalognum.start_end` and `Catalognum.delimited` are always correctly resolved as `re.Pattern`.

Reply:
I think the reason is that the typings for `cached_classproperty` exist in beetcamp, but not in beets ;)

On the other hand, maybe we should simply remove the decorator from `_relation` and turn it into a regular `classmethod`: It's probably too simple to benefit from caching, and in any case, it seems to only be called from methods which in turn are cached using `cached_classproperty`. When I do that, `-> type[Model[D]]` seems to work for mypy.

Reply:
To be honest `cached_` is just a buzzword there, because ultimately it is a _class_ property, and a class is evaluated just once, so 'caching' is kind of inherent here.

The reason why I didn't use a classmethod here comes down to whether the returned data is constant or dynamic. Since it's constant, I thought using a classmethod would be slightly misleading.

Ideally we'd have them set as attributes on the class:

```py
class Album(LibModel):
    _relation = Item
    

class Item(LibModel):
    _relation = Album
```

But you can see this is a chicken and egg problem 😅 since `Item` class is undefined when `Album` class wants to use it. I created this classproperty to work around this limitation, so that they behave as if they were defined as attributes on the class.

Reply:
Yeah, I played around with that code myself, only to run into the same issue 😅

Maybe you could open a quick PR to add typings to `cached_classproperty`? (i.e. just copy them over from beetcamp.)